ARROW-12017: [R] [Documentation] Make proper developing arrow docs (#9898)
apache/arrow: r/vignettes/install.Rmd

after a new release of the `arrow` package will likely fail unless you first
update the system packages.

> ## Using a local Arrow C++ build

Reviewer: Good call, a user should not be here.

## Summary of build environment variables

> Some features are optional when you build Arrow from source

Reviewer: ```suggestion
Some features are optional when you build Arrow from source. With the exception of `ARROW_S3`, these are all `ON` by default in the bundled C++ build, but you can set them to `OFF` to disable them.
```

> * `ARROW_S3`: If set to `true` S3 support will be built (as long as the 
>   dependencies are met, if they are not met, this may be automatically turned off) (default: false)
> * `ARROW_JEMALLOC`: If set to `true` jemalloc support will be included (default: true)
> * `ARROW_PARQUET`: If set to `true` parquet support will be included (default: true)
> * `ARROW_DATASET`: If set to `true` dataset support will be included (default: true)
> * `ARROW_WITH_RE2`: If set to `true` RE2 regular expression library support will be included (default: true)
> * `ARROW_WITH_UTF8PROC`: If set to `true` UTF8Proc string library support will be included (default: true)

Reviewer: ```suggestion
* `ARROW_S3`: If set to `ON` S3 support will be built as long as the 
  dependencies are met; if they are not met, the build script will turn this `OFF` 
* `ARROW_JEMALLOC` for the `jemalloc` memory allocator
* `ARROW_PARQUET`
* `ARROW_DATASET`
* `ARROW_WITH_RE2` for the RE2 regular expression library, used in some string compute functions
* `ARROW_WITH_UTF8PROC` for the UTF8Proc string library, used in many other string compute functions
```

> By default, these are all unset. All boolean variables are case-insensitive.

Reviewer: ```suggestion
There are a number of other variables that affect the `configure` script and the bundled build script.
By default, these are all unset. All boolean variables are case-insensitive.
```

  The directory will be created if it does not exist.
* `CMAKE`: When building the C++ library from source, you can specify a
  `/path/to/cmake` to use a different version than whatever is found on the `$PATH`

> * `ARROW_S3`: If set to `true` S3 support will be built (as long as the 

Reviewer: Let's make this a separate bullet list since these are features to be enabled/disabled in the cmake. Note also what the defaults are. Also might be good to enumerate the compression libraries (could group them all under `ARROW_WITH_*` or something).

Author: I also moved them up to the top of this section, I think these are actually slightly more interesting / important than the other group
